packages/mainmenu-extension/schema/plugin.json

```diff
@@ -78,7 +78,311 @@
       "selector": "body"
     }
   ],
-  "properties": {},
+  "jupyter.lab.transform": true,
+  "properties": {
+    "menus": {
+      "title": "The application menu description.",
+      "description": "Note: To disable a menu or a menu item,\ncopy it to User Preferences and add the\n\"disabled\" key. The following example will disable\nthe \"Tabs\" menu and \"Restart Kernel and Run up to Selected Cell\"\nitem:\n{\n  \"menus\": [\n    {\n      \"id\": \"jp-mainmenu-tabs\",\n      \"disabled\": true\n    },\n    {\n      \"id\": \"jp-mainmenu-kernel\",\n      \"items\": [\n        {\n          \"command\": \"notebook:restart-and-run-to-selected\",\n          \"disabled\": true\n        }\n      ]\n    }\n  ]\n}\n\nMenu description:",
+      "items": {
+        "$ref": "#/definitions/menu"
+      },
+      "type": "array",
+      "default": [
+        {
+          "id": "jp-mainmenu-file",
+          "label": "File",
+          "items": [
```

Comment:
So.... the nested definitions are... not great to edit, and don't give a lot of feedback with the current UI.

Perhaps, regarding the suggestion on the ts side, if the menu rank and the contents were split:

```json
"menuRanks": {
  "jp-mainmenu-file": 1,
  "...": 2,
  "jp-mainmenu-help": 99,
},
"menuItems": {
  "jp-mainmenu-file": [],
}
```

_and_ had an "custom" pattern (a la `extra_labextension_paths` so that it was easy to change _just a little_:

```
"customMenuRanks": {
  "jp-someextension-foo": 10,
  "jp-mainmenu-tabs": null,
}
```

Reply:
Well, shortcuts settings are already nested in very similar matter and it would be consistent to keep it this way. IMO It will not get as nasty as the LSP thing (https://github.com/krassowski/jupyterlab-lsp/pull/418) because it is conceptually easier and much shallower, but also `"additionalProperties": false,` is on and the design expectation is that an average user will not be touching the raw JSON a lot (they will have a dedicated UI for that).

Reply:
I agree with @krassowski 

Small note, the user settings are merged with the default one (taking precedence). Therefore you can minimally change the rank with the following settings (because this is also what the user would like to disable menus for example):

```json
{
    "menus": [
        {
            "id": "jp-mainmenu-file",
            "rank": 50
        },
        {
            "id": "jp-mainmenu-help",
            "rank": 1
        }
    ]
}
```

![image](https://user-images.githubusercontent.com/8435071/120642748-9120dd80-c475-11eb-8a1e-6535e3027de1.png)